api_app/analyzers_manager/file_analyzers/virushee.py

```diff
@@ -0,0 +1,90 @@
+
+import requests
+
+from api_app.analyzers_manager.classes import FileAnalyzer
+from api_app.exceptions import AnalyzerRunException
+from tests.mock_utils import MockResponse, if_mock_connections, patch
+
+logger = logging.getLogger(__name__)
+
+class Virushee(FileAnalyzer):
+    """Analyze a file against Virushee API"""
+
+    def set_params(self, params):
+        self.session = requests.Session()
+        self.max_tries = 30
+        self.poll_distance = 10
+        self.request_url = "https://api.virushee.com/"
+        self.to_force_scan = params.get("force_scan", False)
+
+    def run(self):
+        binary = self.read_file_bytes()
+        if not binary:
+            raise AnalyzerRunException("File is empty")
+        if not self.to_force_scan:
+            hash_result = self.__check_report_for_hash()
+            if hash_result:
+                return hash_result
+        task_id = self.__upload_file(binary)
+        result = self.__poll_status(task_id)
+        return result
+
+    def __check_report_for_hash(self):
+        api_url = f"{self.request_url}file/hash/{self.md5}"
+        try:
+            response = requests.get(api_url)
+            response.raise_for_status()
+        except:
+            return None
+
+    def __upload_file(self, binary) -> str:
+        name_to_send = self.filename if self.filename else self.md5
+        files = {"file": (name_to_send, binary)}
+        upload_url = self.request_url + "file/upload"
+        try:
+            response = self.session.post(upload_url, files=files)
+            response.raise_for_status()
+        except requests.RequestException as error:
+            raise AnalyzerRunException(error)
+        return response.json()["task"]
+
+    def __poll_status(self, task_id: str) -> dict:
+        for chance in range(self.max_tries):
+            logger.info(f"Polling #try{chance+1}")
+            time.sleep(self.poll_distance)
+            request_url = self.request_url + "file/task/" + task_id
+            try:
+                response = self.session.get(request_url)
+                response.raise_for_status()
+            except:
+                continue
+            response = self.session.get(request_url)
+            if response.status_code == 422:
+                raise AnalyzerRunException(
+                    "Virushee API returned an error: " + response.text
+                )
+            if response.status_code == 200:
+                return response.json()
+
+    @classmethod
+    def _monkeypatch(cls):
+        patches = [
+            if_mock_connections(
+                patch(
+                    "requests.Session.get",
+                    return_value=MockResponse({"message": "hash_found"}, 200),
+                ),
+                patch(
+                    "requests.Session.post",
+                    return_value=MockResponse(
+                        {"task": "80ca33ee-2df2-489a-9444-886db9abc5f0"}, 201
+                    ),
+                ),
```

Comment:
missing the patch statement for the GET request being sent inside `__poll_status` method.

Reply:
I'm not sure what to do here as I'm not much acquainted with this patching thng. However, I've added an extra mocck response for `https://api.virushee.com/file/task/{id}`
is it fine?

Reply:
we are monkeypatch-ing these requests to mock them (i.e. actual HTTP request is not sent and only the `return_value` is returned). So you need to define patches in order of the requests being made by the analyzer code. In the normal case, there are 3 requests being made: check hash GET -> upload file POST -> poll status GET, so 3 patch statements are required here.

Reply:
We want to test maximum number of lines of the code, so for that we need to mimic the case of hash_not_found. i.e.,

there are 4 requests being made: 
- GET check_hash, 404 in response
- POST upload_file, 201 in response with task_id
- GET poll_result 202 (analysis still running)
- GET poll_result 200 (got result)

so total 4 `patch` statements required.

Reply:
Thanks for the detailed explanation. Understood!!